sync/fromDisk/questions.js

```diff
@@ -1,124 +1,115 @@
+                    const directories = questions.map(q => q.directory).join(', ');
+                    throw new Error(`UUID ${uuid} is used in multiple questions: ${directories}`);
+                }
+            });
+
+        // Preprocess questions into obects before sending them to the sproc
+        // for syncing
+        const questionsParam = Object.keys(questionDB).map(qid => {
+            const q = questionDB[qid];
+
+            let external_grading_files = null;
+            if (q.externalGradingOptions) {
+                const opts = q.externalGradingOptions;
+                if (opts.files && opts.serverFilesCourse) {
+                    throw new Error(`Question ${qid} cannot use both externalGradingOptions.files and externalGradingOptions.serverFilesCourse`);
+                } else if (opts.files) {
+                    jobLogger.warn(`WARNING: Question ${qid} uses externalGradingOptions.files, which will be deprecated in favor of externalGradingOptions.serverFilesCourse`);
+                    external_grading_files = opts.files;
+                } else if (opts.serverFilesCourse) {
+                    external_grading_files = opts.serverFilesCourse;
+                }
+            }
+
+            let partialCredit;
+            if (q.partialCredit != null) {
+                partialCredit = q.partialCredit;
+            } else {
+                if (q.type == 'v3') {
+                    partialCredit = true;
+                } else {
+                    partialCredit = false;
+                }
+            }
+            return {
+                uuid: q.uuid,
+                qid: qid,
+                type: (q.type == 'v3') ? 'Freeform' : q.type,
+                title: q.title,
+                partial_credit: partialCredit,
+                template_directory: q.template,
+                options: q.options,
+                client_files: q.clientFiles || [],
+                topic: q.topic,
+                grading_method: q.gradingMethod || 'Internal',
+                single_variant: !!q.singleVariant,
+                external_grading_enabled: (q.externalGradingOptions && q.externalGradingOptions.enabled),
+                external_grading_image: (q.externalGradingOptions && q.externalGradingOptions.image),
+                external_grading_files: external_grading_files,
+                external_grading_entrypoint: (q.externalGradingOptions && q.externalGradingOptions.entrypoint),
+                external_grading_timeout: (q.externalGradingOptions && q.externalGradingOptions.timeout),
+                external_grading_enable_networking: (q.externalGradingOptions && q.externalGradingOptions.enableNetworking),
+            };
         });
-    },
-};
+
+        // Sync the questions to the DB. This will create/update all questions,
+        // soft-delete any unused questions, and ensure that all questions have
+        // assigned numbers (TODO clarify what these numbers are?)
```

Comment:
These are the numbers referred to here: https://prairielearn.readthedocs.io/en/latest/assessment/#changing-question-order-randomization